src/Std/Data/TreeMap/RawLemmas.lean

```diff
@@ -47,4 +50,171 @@ theorem isEmpty_insertIfNew [TransCmp cmp] (h : t.WF) {k : α} {v : β} :
+
+theorem isEmpty_eq_false_iff_exists_mem [TransCmp cmp] (h : t.WF) :
+    t.isEmpty = false ↔ ∃ a, a ∈ t :=
+  DTreeMap.Raw.isEmpty_eq_false_iff_exists_mem h
+
+theorem isEmpty_iff_forall_contains [TransCmp cmp] (h : t.WF) :
+    t.isEmpty = true ↔ ∀ a, t.contains a = false :=
+  DTreeMap.Raw.isEmpty_iff_forall_contains h
+
+theorem isEmpty_iff_forall_not_mem [TransCmp cmp] (h : t.WF) :
+    t.isEmpty = true ↔ ∀ a, ¬a ∈ t :=
+  DTreeMap.Raw.isEmpty_iff_forall_not_mem h
+
+@[simp]
+theorem insert_eq_insert {p : α × β} : Insert.insert p t = t.insert p.1 p.2 :=
+  rfl
+
+@[simp]
+theorem singleton_eq_insert {p : α × β} :
+    Singleton.singleton p = (∅ : Raw α β cmp).insert p.1 p.2 :=
+  rfl
+
+@[simp]
+theorem contains_insert [h : TransCmp cmp] (h : t.WF) {k a : α} {v : β} :
+    (t.insert k v).contains a = (cmp k a == .eq || t.contains a) :=
+  DTreeMap.Raw.contains_insert h
+
+@[simp]
+theorem mem_insert [TransCmp cmp] (h : t.WF) {k a : α} {v : β} :
+    a ∈ t.insert k v ↔ cmp k a = .eq ∨ a ∈ t :=
+  DTreeMap.Raw.mem_insert h
+
+theorem contains_insert_self [TransCmp cmp] (h : t.WF) {k : α} {v : β} :
+    (t.insert k v).contains k :=
+  DTreeMap.Raw.contains_insert_self h
+
+theorem mem_insert_self [TransCmp cmp] (h : t.WF) {k : α} {v : β} :
+    k ∈ t.insert k v :=
+  DTreeMap.Raw.mem_insert_self h
+
+theorem contains_of_contains_insert [TransCmp cmp] (h : t.WF) {k a : α} {v : β} :
+    (t.insert k v).contains a → ¬ cmp k a = .eq → t.contains a :=
+  DTreeMap.Raw.contains_of_contains_insert h
+
+theorem mem_of_mem_insert [TransCmp cmp] (h : t.WF) {k a : α} {v : β} :
+    a ∈ t.insert k v → ¬ cmp k a = .eq → a ∈ t :=
+  DTreeMap.Raw.mem_of_mem_insert h
+
+@[simp]
+theorem size_emptyc : (∅ : Raw α β cmp).size = 0 :=
+  DTreeMap.Raw.size_emptyc
+
+theorem isEmpty_eq_size_eq_zero (h : t.WF) :
+    t.isEmpty = (t.size == 0) :=
+  DTreeMap.Raw.isEmpty_eq_size_eq_zero h.out
+
+theorem size_insert [TransCmp cmp] (h : t.WF) {k : α} {v : β} :
+    (t.insert k v).size = if t.contains k then t.size else t.size + 1 :=
+  DTreeMap.Raw.size_insert h
+
+theorem size_le_size_insert [TransCmp cmp] (h : t.WF) {k : α} {v : β} :
+    t.size ≤ (t.insert k v).size :=
+  DTreeMap.Raw.size_le_size_insert h
+
+theorem size_insert_le [TransCmp cmp] (h : t.WF) {k : α} {v : β} :
+    (t.insert k v).size ≤ t.size + 1 :=
+  DTreeMap.Raw.size_insert_le h
+
+@[simp]
+theorem erase_emptyc {k : α} :
+    (empty : Raw α β cmp).erase k = empty :=
+  ext <| DTreeMap.Raw.erase_emptyc
+
+@[simp]
+theorem isEmpty_erase [TransCmp cmp] (h : t.WF) {k : α} :
+    (t.erase k).isEmpty = (t.isEmpty || (t.size == 1 && t.contains k)) :=
+  DTreeMap.Raw.isEmpty_erase h
+
+@[simp]
+theorem contains_erase [TransCmp cmp] (h : t.WF) {k a : α} :
+    (t.erase k).contains a = (cmp k a != .eq && t.contains a) :=
+  DTreeMap.Raw.contains_erase h
+
+@[simp]
+theorem mem_erase [TransCmp cmp] (h : t.WF) {k a : α} :
+    a ∈ t.erase k ↔ ¬ cmp k a = .eq ∧ a ∈ t :=
```

Comment:
I think we generally state things in terms of `≠` rather than `¬ =` even though this is not simp-normal form. `ne_eq` really is a truly strange phenomenon.